include/deal.II/differentiation/ad/ad_number_traits.h

```diff
@@ -0,0 +1,1043 @@
+       * the real-type equivalent that its associated with. For Sacado, and
+       * likely other AD number types,  the floating point value stored in
+       * an @p ADNumberType must be extracted through some function that is
+       * specific to each type of AD number. This requires some specialist
+       * intervention to get at this data.
+       *
+       * @author Jean-Paul Pelteret, 2017
+       */
+      template <typename T>
+      struct NumberType;
+
+
+      namespace
+      {
+        /**
+         * A small struct to remove the @p std::complex wrapper
+         * around a number.
+         *
+         * @author Jean-Paul Pelteret, 2017
+         */
+        template<typename Number>
+        struct RemoveComplexWrapper;
+      }
+
+    } // namespace internal
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a supported
+     * auto-differentiable number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename NumberType>
+    struct is_ad_number;
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a taped
+     * auto-differentiable number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename NumberType, typename = void>
+    struct is_taped_ad_number;
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a tapeless
+     * auto-differentiable number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename NumberType, typename = void>
+    struct is_tapeless_ad_number;
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a real-valued
+     * auto-differentiable number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename NumberType, typename = void>
+    struct is_real_valued_ad_number;
+
+
+    /**
+     * A struct to indicate whether a given @p NumberType is a complex-valued
+     * auto-differentiable number or not. By default, numbers are not
+     * considered to have the necessary characteristics to fulfill this
+     * condition.
+     *
+     * @author Jean-Paul Pelteret, 2017
+     */
+    template <typename NumberType, typename = void>
+    struct is_complex_valued_ad_number;
+
+  } // namespace AD
+} // namespace Differentiation
+
+
+/* --------------------------- inline and template functions and specializations ------------------------- */
+
+
+#ifndef DOXYGEN
```

Comment:
may I ask a silly question? why there is this guard?

Reply:
Not a silly question at all :-) All of the specialisations (e.g. [this one](https://github.com/jppelteret/dealii/blob/5951012c61c1603f0037a3f99fc9828b71d461a3/include/deal.II/differentiation/ad/ad_number_traits.h#L728)) completely pollute the documentation, in particular the [class index](http://dealii.org/developer/doxygen/deal.II/classes.html). I'd happily have left them in if it weren't for the fact that the columns end up being more than a screen wide!

Reply:
My compromise was to comprehensively document these template classes / functions once off and to have less detailed internal documentation within the guards.

Reply:
I see. thanks